src/main/scala/com/example/App.scala

```diff
@@ -1,40 +1,53 @@
-import com.example.reporters._
-import com.example.reporters.common.util.HtmlUtil
+import com.example.model.{Answer, Question}
 import com.example.session.SparkSessionHolder
 import com.example.util.CommonUtil
+import org.apache.log4j.Logger
+import org.apache.spark.ml.evaluation.RegressionEvaluator
+import org.apache.spark.ml.regression.{GBTRegressor, LinearRegression}
+import org.apache.spark.ml.tuning.{CrossValidatorModel, ParamGridBuilder}
+import org.apache.spark.sql.DataFrame
 
 object App {
+  private val log = Logger.getLogger(App.getClass)
+
   def init(args: Array[String]): Unit = {
     CommonUtil.baseFileLocation(args(0))
     SparkSessionHolder.init(args(1), args(2))
   }
 
+  def logModelMetrics(model: CrossValidatorModel, testData: DataFrame): Unit = {
+    val predictions = model.transform(testData)
+    val rmse = new RegressionEvaluator().setMetricName("rmse").evaluate(predictions)
+    val mae = new RegressionEvaluator().setMetricName("mae").evaluate(predictions)
+
+    log.info("**********************************")
+    log.info(s"rmse: $rmse")
+    log.info(s"mae: $mae")
+    log.info("**********************************")
+  }
+
   def main(args: Array[String]): Unit = {
     init(args)
-
-    val questionsDs = Extractor.readCsv[Question]("Questions.csv")
     val answersDs = Extractor.readCsv[Answer]("Answers.csv")
-    val tagsDs = Extractor.readCsv[Tag]("Tags.csv")
-
-    val majorTopicsReporter = MajorTopicsReporter(tagsDs, 100)
-    val expertsReporter = ExpertsReporter(questionsDs, answersDs, tagsDs, majorTopicsReporter.reportData.df)
-    val reporters = Seq(
-      DailyStatisticsReporter(questionsDs, answersDs),
-      AverageScoreReporter(answersDs, (a: Answer) => HtmlUtil.containsLinks(a.body), "avg_contains_link_score"),
-      AverageScoreReporter(answersDs, (a: Answer) => !HtmlUtil.containsLinks(a.body), "avg_not_contains_link_score"),
-      AverageScoreReporter(answersDs, (a: Answer) => HtmlUtil.containsCode(a.body), "avg_contains_code_score"),
-      AverageScoreReporter(answersDs, (a: Answer) => !HtmlUtil.containsCode(a.body), "avg_not_contains_code_score"),
-      CodeLenReporter(answersDs, questionsDs),
-      majorTopicsReporter,
-      expertsReporter,
-      QuotedReporter(answersDs, 100),
-      SeasonalityReporter(questionsDs, answersDs)
-    )
-
-    reporters.foreach(_.report)
+    val questionsDs = Extractor.readCsv[Question]("Questions.csv")
+
+    val data = AnswerScoreModel.transformData(answersDs, questionsDs)
+    val Array(train, test) = data.randomSplit(Array(.8, .2))
+
+    val linearRegression = new LinearRegression()
+    val linearParams = new ParamGridBuilder().addGrid(linearRegression.elasticNetParam, Array(0.3, 0.6, 0.8)).build()
+    val linearModel = AnswerScoreModel.model(train, linearParams, linearRegression)
+
+
+    val gbtRegressor = new GBTRegressor()
+    val gbtParams = new ParamGridBuilder().addGrid(gbtRegressor.featureSubsetStrategy, Array("auto", "all")).build()
```

Comment:
very few params tuned here. Please see the attached learning notebook you can use as a reference.

[ML 08 - Tuning Bikeshare.scala.zip](https://github.com/temp-sprk-lrng/spark-lrng/files/3780706/ML.08.-.Tuning.Bikeshare.scala.zip)